src/client/font_enriched_string.cpp

```diff
@@ -0,0 +1,205 @@
+// Luanti
+// SPDX-License-Identifier: LGPL-2.1-or-later
+// Copyright (C) 2024 cx384
+
+#include "font_enriched_string.h"
+#include "irrlicht_changes/CGUITTFont.h"
+#include "util/string.h"
+#include <utility>
+
+enum class FontModifier : u8
+{
+	Mono,
+	Unmono,
+	Bold,
+	Unbold,
+	Italic,
+	Unitalic,
+};
+
+static void applyFontModifier(FontSpec &spec, FontModifier modifier)
+{
+	switch (modifier) {
+		case FontModifier::Mono :
+			spec.mode = FM_Mono;
+			break;
+		case FontModifier::Unmono :
+			spec.mode = FM_Standard;
+			break;
+		case FontModifier::Bold :
+			spec.bold = true;
+			break;
+		case FontModifier::Unbold :
+			spec.bold = false;
+			break;
+		case FontModifier::Italic :
+			spec.italic = true;
+			break;
+		case FontModifier::Unitalic :
+			spec.italic = false;
+			break;
+		default:
+			break;
+	}
+}
+
+static bool parseFontModifier(std::string_view s, FontModifier &modifier)
+{
+	if (s.empty())
+		return false;
+
+	switch (s[0]) {
+		case 'M':
+			modifier = FontModifier::Mono;
+			break;
+		case 'm':
+			modifier = FontModifier::Unmono;
+			break;
+		case 'B':
+			modifier = FontModifier::Bold;
+			break;
+		case 'b':
+			modifier = FontModifier::Unbold;
+			break;
+		case 'I':
+			modifier = FontModifier::Italic;
+			break;
+		case 'i':
+			modifier = FontModifier::Unitalic;
+			break;
+		default:
+			return false;
+	}
+	return true;
+};
+
+FontEnrichedString::FontEnrichedString(const std::wstring &s,
+		video::SColor initial_color, const FontSpec &initial_font)
+{
+
+	FontSpec font = initial_font;
+	video::SColor color = initial_color;
+
+	// Handle font escape sequence like in EnrichedString and translate_string
+	Line line;
+	size_t fragmen_start = 0;
+	size_t i = 0;
+	while (i < s.length()) {
+		if (s[i] == L'\n') {
+			// Split lines
+
+			EnrichedString fragment(std::wstring(s, fragmen_start, i - fragmen_start), color);
+			auto colors = fragment.getColors();
+			line.emplace_back(std::make_pair(std::move(fragment), font));
```

Comment:
the pair will be constructed inline, that's the point
```suggestion
			line.emplace_back(std::move(fragment), font);
```